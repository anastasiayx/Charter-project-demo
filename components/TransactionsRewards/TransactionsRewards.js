import React from "react";
import { fetchTransactions } from '../../services/transactions.service';
import { Collapse, Typography } from 'antd';
import CustomerRewards from '../CustomerRewards/CustomerRewards'
import { arrayGroupByPropertyValue, sumArrayItemPropertyNumValue } from '../../utils/array.helper'
import { TRANSACTIONS } from '../../constants/transactions.constant';
import { useTransactions } from '../../hooks/useTransactions'




const TransactionRewards = () => {
    const [customerTransactionsMap, setcustomerTransactionsMap] = React.useState({});
    const [res, isLoading, eror] = useTransactions()
    React.useEffect(() => {
        if (res?.data?.resultData) {
            const customerMap = arrayGroupByPropertyValue(res.data.resultData, 'c_id')
            setcustomerTransactionsMap(customerMap)
        }
    }, [res])

    const renderCustomerRewards = React.useMemo(() => {
        return <Collapse defaultActiveKey="panel-customer-0">
            {
                Object.entries(customerTransactionsMap).map(([customerId, customerTransactions], index) => {
                    return <Collapse.Panel header={`Customer ID:${customerId}, Total Rewards:${sumArrayItemPropertyNumValue(customerTransactions, 'rewards')}`} key={`panel-customer-${index}`}>
                        <CustomerRewards transactions={customerTransactions} />
                    </Collapse.Panel>
                })
            }
        </Collapse>

    }, [customerTransactionsMap])


    return <section>
        <header>
            <Typography.Title role="heading">{TRANSACTIONS.PORTAL_TITLE}</Typography.Title>
        </header>

        {
            isLoading ? <span className="loading-spinner">Loading...</span> : renderCustomerRewards
        }

    </section>;
}

export default TransactionRewards