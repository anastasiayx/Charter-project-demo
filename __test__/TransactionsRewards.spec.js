import { render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import TransactionsRewards from '@/components/TransactionsRewards/TransactionsRewards'
import { setupServer } from 'msw/node'
import { rest } from 'msw'
import './matchMediaMock'


describe('TransactionsRewards', () => {
    const transactionsMock = [
        {
            t_id: "1",
            spent: 130,
            create_date: new Date("8/29/22").toISOString(),
            rewards: 110,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "2",
            spent: 200,
            create_date: new Date("7/29/22").toISOString(),
            rewards: 250,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "3",
            spent: 30,
            create_date: new Date("6/29/22").toISOString(),
            rewards: 0,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "4",
            spent: 187,
            create_date: new Date("7/29/22").toISOString(),
            rewards: 224,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "5",
            spent: 130,
            create_date: new Date("8/29/22").toISOString(),
            rewards: 110,
            c_id: "asdfdsffddsf"
        },
        {
            t_id: "6",
            spent: 200,
            create_date: new Date("7/29/22").toISOString(),
            rewards: 250,
            c_id: "asdfdsffddsf"
        },
        {
            t_id: "7",
            spent: 154,
            create_date: new Date("7/19/22").toISOString(),
            rewards: 158,
            c_id: "asdfdsffddsf"
        },
        {
            t_id: "8",
            spent: 43,
            create_date: new Date("7/29/22").toISOString(),
            rewards: 0,
            c_id: "asdfdsffddsf"
        }, {
            t_id: "9",
            spent: 187,
            create_date: new Date("8/29/22").toISOString(),
            rewards: 224,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "10",
            spent: 130,
            create_date: new Date("6/24/22").toISOString(),
            rewards: 110,
            c_id: "asdfdsffddsf"
        },
        {
            t_id: "11",
            spent: 50,
            create_date: new Date("8/19/22").toISOString(),
            rewards: 0,
            c_id: "asdfdsffddsf"
        },
        {
            t_id: "12",
            spent: 54,
            create_date: new Date("7/26/22").toISOString(),
            rewards: 4,
            c_id: "asdfdsffddsf"
        },
        {
            t_id: "13",
            spent: 43,
            create_date: new Date("7/01/22").toISOString(),
            rewards: 0,
            c_id: "asdfdsffddsf"
        }
    ]

    const server = setupServer(
        rest.get('/api/transactions', (req, res, ctx) => {
            return res(ctx.json({ resultData: transactionsMock }))
        }),
    )

    beforeAll(() => server.listen())
    afterEach(() => server.resetHandlers())
    afterAll(() => server.close())


    it(' shoud match the snapshot', async () => {
        const { asFragment, queryByText } = render(<TransactionsRewards />)
        expect(asFragment()).toMatchSnapshot()
        await waitForElementToBeRemoved(() => queryByText('Loading...'));
    })


    it('renders a Transactions Rewards Portal heading', async () => {
        const { queryByText } = render(<TransactionsRewards />)
        const heading = screen.getByRole('heading', {
            name: /Transactions Rewards Portal/i,
        })
        expect(heading).toBeInTheDocument()
        await waitForElementToBeRemoved(() => queryByText('Loading...'));

    })

    it('will show customer panels after page has loaded the transaction data', async () => {
        const { queryByText } = render(<TransactionsRewards />)
        let loading = queryByText('Loading...')
        expect(loading).toBeInTheDocument()
        await waitForElementToBeRemoved(() => queryByText('Loading...'));

        expect(queryByText('Customer ID:sdfs34234fss, Total Rewards:808')).toBeInTheDocument()
        expect(queryByText('Customer ID:asdfdsffddsf, Total Rewards:632')).toBeInTheDocument()
    })
})