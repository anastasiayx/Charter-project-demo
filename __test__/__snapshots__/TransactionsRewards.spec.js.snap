// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsRewards  shoud match the snapshot 1`] = `
<DocumentFragment>
  <section>
    <header>
      <h1
        class="ant-typography"
        role="heading"
      >
        Transactions Rewards Portal
      </h1>
    </header>
    <span
      class="loading-spinner"
    >
      Loading...
    </span>
  </section>
</DocumentFragment>
`;
